src/support/qsort.c

```diff
@@ -0,0 +1,269 @@
+    }
+}
+
+/*
+ * __med3 --
+ *     Find the median of three values, using the user-supplied comparator.
+ */
+static inline void *
+__med3(void *a, void *b, void *c, wt_cmp_t cmp, void *context)
+{
+#ifndef WT_QSORT_R_DEF
+    WT_UNUSED(context);
+#endif
+
+    if (CMP(a, b, context) < 0) {
+        if (CMP(b, c, context) < 0)
+            return (b);
+        else
+            return (CMP(a, c, context) < 0 ? c : a);
+    } else {
+        if (CMP(b, c, context) > 0)
+            return (b);
+        else
+            return (CMP(a, c, context) < 0 ? a : c);
+    }
+}
+
+/*
+ * __qsort --
+ *     Sort an array using a user-supplied comparator function and a context argument.
+ */
+static void
+__qsort(void *arr, size_t nmemb, size_t elem_sz, wt_cmp_t cmp, void *context)
+{
+    /*
+     * This is originally from "Engineering a Sort Function", by Jon L. Bentley and Douglas
+     * M. McIlroy ("Software - Practice and Experience, Vol. 23(11)", November 1993). It's been
+     * massaged into a form suitable for WiredTiger, initially by Luke Pearson.
+     *
+     * It's not simple, so some description of where all these variables point is in order:
+     * -----------------------------------------------
+     * | = |        <         | ? |      >       | = |
+     * -----------------------------------------------
+     *      ^                  ^ ^              ^
+     *      |                  | |              |
+     *      lowest_lt_median   | hi_unknown     highest_gt_median
+     *                         lo_unknown
+     */
+    bool swapped;
+    int cmp_result;
+    size_t lhs_unsorted, rhs_unsorted;
+    uint8_t *a, *hi_pseudomedian, *hi_unknown, *highest_gt_median, *lo_pseudomedian, *lo_unknown,
+      *lowest_lt_median, *pseudomedian;
+
+    WT_ASSERT(NULL, cmp != NULL);
+    if (UNLIKELY(nmemb < 2))
+        return;
+
+    a = arr; /* Avoid arithmetic on void pointers. */
+
+    for (;;) {
+        swapped = false;
+
+        /* Insertion sort for small arrays. */
+        if (nmemb < WT_QSORT_INSERT_THRESHOLD) {
+            for (hi_unknown = a + elem_sz; hi_unknown < a + nmemb * elem_sz; hi_unknown += elem_sz)
+                for (lo_unknown = hi_unknown;
+                     lo_unknown > a && CMP(lo_unknown - elem_sz, lo_unknown, context) > 0;
+                     lo_unknown -= elem_sz)
+                    __swap_bytes(lo_unknown, lo_unknown - elem_sz, elem_sz);
+            return;
+        }
+
+        /* Take middle element as median for small arrays. */
+        pseudomedian = a + (nmemb / 2) * elem_sz;
+        if (nmemb > WT_QSORT_INSERT_THRESHOLD) {
+            lo_pseudomedian = a;
+            hi_pseudomedian = a + (nmemb - 1) * elem_sz;
+            /*
+             * For larger arrays, do more work to find a good median - here, it's a "pseudo-median"
+             * of nine samples.
+             */
+            if (nmemb > WT_QSORT_LARGE_THRESHOLD) {
+                size_t median_dist = (nmemb / 8) * elem_sz;
```

Comment:
Afaict the `8` here is dependent on the value of `WT_QSORT_LARGE_THRESHOLD`. This is not obvious, and will be suboptimal if the define changes. I guess it could break, but it seems unlikely the value will decrease.

Reply:
Nope, the eight is just "a reasonable distance into the array" - it could just as easily be 7 or 9. The idea is that, in this "median of medians" code, you take samples 1/8th of the array size past the beginning, 2/8th past the beginning, etc. That said, if it wasn't obvious it needs a comment. Did the above explanation help? If so I'll base the comment on that.

Reply:
If I understand correctly looking back at the code you sample 9 roughly equidistant elements within the range?

Reply:
Yes, that's right.